compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -723,6 +723,138 @@ OMR::ARM64::TreeEvaluator::lremEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    return iremHelper(node, true, cg);
    }
 
+/**
+ * @brief Generates ubfm instruction for mask and shift operation if possible
+ *
+ * @param[in] shiftNode: node with shift opcode
+ * @param[in]        cg: code generator
+ *
+ * @returns register which contains the result of the operation. NULL if the operation cannot be encoded in ubfm instruction.
+ */
+static TR::Register *
+generateUBFMForMaskAndShift(TR::Node *shiftNode, TR::CodeGenerator *cg)
+   {
+   TR::Node *andNode = NULL;
+   TR::Node *shiftValueNode = NULL;
+   TR::Node *maskNode = NULL;
+   TR::Node *sourceNode = NULL;
+   if (shiftNode->getFirstChild()->getOpCode().isAnd())
+      {
+      andNode = shiftNode->getFirstChild();
+      shiftValueNode = shiftNode->getSecondChild();
+      if (!shiftValueNode->getOpCode().isLoadConst())
+         {
+         return NULL;
+         }
+      if ((andNode->getReferenceCount() > 1) || (andNode->getRegister() != NULL))
+         {
+         return NULL;
+         }
+      if (andNode->getSecondChild()->getOpCode().isLoadConst())
+         {
+         sourceNode = andNode->getFirstChild();
+         maskNode = andNode->getSecondChild();
+         }
+      else if (andNode->getFirstChild()->getOpCode().isLoadConst())
+         {
+         sourceNode = andNode->getSecondChild();
+         maskNode = andNode->getFirstChild();
+         }
+      }
+
+   if (maskNode == NULL)
+      {
+      return NULL;
+      }
+   const bool is64bit = shiftNode->getDataType().isInt64();
+   const int64_t shiftValue = shiftValueNode->getConstValue();
+
+   if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+      {
+      return NULL;
+      }
+   const uint64_t maskValue = is64bit ? maskNode->getLongInt() : static_cast<uint32_t>(maskNode->getInt());
```

Comment:
I suppose you mean `static_cast<uint64_t>`.

Reply:
Actually, we need a cast to `uint32_t` to prevent sign extension to make it work when the value of maskNode is negative.

Reply:
I see.  Thanks.